python/ray/utils.py

```diff
@@ -391,15 +391,43 @@ def setup_logger(logging_level, logging_format):
     logger.propagate = False
 
 
+class Unbuffered(object):
+    """There's no "built-in" solution to programatically disabling buffering of
+    text files. Ray expects stdout/err to be text files, so creating an
+    unbuffered binary file is unacceptable.
+
+    See
+    https://mail.python.org/pipermail/tutor/2003-November/026645.html.
+    https://docs.python.org/3/library/functions.html#open
+
+    """
+
+    def __init__(self, stream):
+        self.stream = stream
+
+    def write(self, data):
+        self.stream.write(data)
+        self.stream.flush()
+
+    def writelines(self, datas):
+        self.stream.writelines(datas)
+        self.stream.flush()
+
+    def __getattr__(self, attr):
+        return getattr(self.stream, attr)
+
+
 def open_log(path, **kwargs):
     """
     Opens the log file at `path`, with the provided kwargs being given to
     `open`.
     """
+    # Disable buffering, see test_advanced_3.py::test_logging_to_driver
     kwargs.setdefault("buffering", 1)
     kwargs.setdefault("mode", "a")
     kwargs.setdefault("encoding", "utf-8")
-    return open(path, **kwargs)
+    stream = open(path, **kwargs)
+    return Unbuffered(stream)
```

Comment:
I don't think we want all of the ray logs to be unbuffered and I think this is used in the non-testing path. Can we just do something like: `print("Hello world!", flush=True)`?

Reply:
I think we want to avoid the user having to supply `flush=True` but yeah, we can make this happen for stdout/err only. In practice it shouldn't make a difference since the default is line buffering and all logs should be written at least one line at a time.